compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -1030,6 +1031,36 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let t_expr = self.check_expr_with_expectation(e, ExpectCastableToType(t_cast));
         let t_cast = self.resolve_vars_if_possible(t_cast);
 
+        // ptr_to_int_cast lint: when casting a function pointer to a numeric type it only makes
+        // sense to cast to `usize`. Other cases are often bugs. See #81686 as an example.
+        if t_expr.is_fn() || t_expr.is_raw_ptr() {
+            let to_numeric = matches!(t_cast.kind(), ty::Int(_) | ty::Uint(_) | ty::Float(_));
+            if to_numeric && !matches!(t_cast.kind(), ty::Uint(ty::UintTy::Usize)) {
+                self.tcx().struct_span_lint_hir(
+                    lint::builtin::PTR_TO_INT_CAST,
+                    expr.hir_id,
+                    expr.span,
+                    |lint| {
+                        let msg = format!("casting pointer to {}", t_cast);
+                        lint.build(&msg)
+                            .span_label(expr.span, &msg)
+                            .span_label(
```

Comment:
```suggestion
                            .help(
```
I think this works for lints too, and it's more appropriate than a label if so.